### ISSUE TITLE
Triangulation error on tilted grids.

### ISSUE THREAD

**Issue by MK-3PP** [abellgithub/delaunator-cpp#8]
I am processing point clouds around the size of 100 k to 1.000 k Points. The points' outer hull is a near perfect rectangle and their distribution is grid like. In pre-processing the point cloud gets slightly tilted towards a plane for error compensation.

While I have not yet observed the issue below in our raw data, occasionally there was a number of big triangles (hundreds in a cloud of a million points) spanning across the whole pre-processed (tilted) point cloud. To reproduce this, I wrote a test. It generates simple square grid, rotates it by a few degrees and then calls delaunator to triangulate.
Validation is done knowing the grid consists of 1-by-1 unit squares, thus the triangulation should consist of right triangles with two unit-length edges and one sqrt(2) unit-length edge. Degenerate triangles on the outer hull (due to point rotation rounding errors) are discarded by checking for very small triangle areas.

The smallest point set I could find to let the test fail is the one supplied below.
Compiler is Visual C++ 14.0, tested platforms are Win32 and x64. Testing framework is googletest, but you should have no difficulties to modify it for Catch (I did not because I do not fully understand the test structure in your repo).

A plot of the triangulated grid is attached here:
![rotate_grid_6_8_8_207_0](https://user-images.githubusercontent.com/60696116/77438210-d4011300-6de5-11ea-8837-dfc1135fa1ed.png)

```
#include "gtest/gtest.h"
#include "delaunator/delaunator.hpp"

struct Point2D {
    double x, y;

    Point2D()
        : x(std::numeric_limits<double>::quiet_NaN())
        , y(std::numeric_limits<double>::quiet_NaN())
    {}

    Point2D(double x, double y)
        : x(x)
        , y(y)
    {}
};

Point2D rotate(Point2D const& p_in, Point2D const& center, double const& angle_deg) {
    const double pi        = 3.14159265358979323846;
    const double angle_rad = angle_deg * pi / 180.0;
    const double s         = sin(angle_rad);
    const double c         = cos(angle_rad);
    Point2D p;

    // translate point back to origin:
    p.x = p_in.x - center.x;
    p.y = p_in.y - center.y;

    // rotate point
    double tmp = p.x * c - p.y * s;
    p.y = p.x * s + p.y * c;
    p.x = tmp;

    // translate point back:
    p.x += center.x;
    p.y += center.y;

    return p;
}

double distance(Point2D const& p1, Point2D const& p2) {
    return sqrt((p1.x - p2.x) * (p1.x - p2.x) + (p1.y - p2.y) * (p1.y - p2.y));
}

bool approx(double v1, double v2) {
    const double eps = 0.01;
    return fabs(v1 - v2) < eps;
}

TEST(External_Lib_Delaunator_Test, rotate_grid_6_8_8_207_0) {
    const size_t  grid_size = 6;
    const Point2D center(8, 8);
    const double  angle = 207.0;

    std::vector<double> points;

    // Generate rotated point grid of 1-by-1 squares.
    for (size_t x = 0; x < grid_size; ++x) {
        for (size_t y = 0; y < grid_size; ++y) {
            Point2D p = rotate(Point2D(x, y), center, angle);
            points.push_back(p.x);
            points.push_back(p.y);
        }
    }

    delaunator::Delaunator delaunator(points);

    // Validate.
    // Assume:
    // (1) since the grid is 1-by-1 square, that each triangle has to have two edges with length 1 and one edge wit length sqrt(2) - give or take epsilon.
    // (2) due to rounding errors during rotation the hull points of the grid are not in a numerically perfect line, thus degenerate triangles will appear on the outer edges. Their area mitght be rather small (< 0.01) compared to a regular grid triangle (== 0.5).
    for (size_t i = 0; i < delaunator.triangles.size(); i += 3) {
        Point2D p1(points[2 * delaunator.triangles[i]],     points[1 + 2 * delaunator.triangles[i]]);
        Point2D p2(points[2 * delaunator.triangles[i + 1]], points[1 + 2 * delaunator.triangles[i + 1]]);
        Point2D p3(points[2 * delaunator.triangles[i + 2]], points[1 + 2 * delaunator.triangles[i + 2]]);

        // Determine edge lengths and triangle area.
        double len_1 = distance(p1, p2);
        double len_2 = distance(p2, p3);
        double len_3 = distance(p3, p1);
        double area  = fabs((p1.x * (p2.y - p3.y) + p2.x * (p3.y - p1.y) + p3.x * (p1.y - p2.y)) / 2.0);

        // Check edge lengths according to (1).
        bool cond_1 = false;
        if (approx(len_1, sqrt(2.0))) {
            cond_1 = approx(len_2, 1.0) && approx(len_3, 1.0);
        }
        else if (approx(len_2, sqrt(2.0))) {
            cond_1 = approx(len_1, 1.0) && approx(len_3, 1.0);
        }
        else if (approx(len_3, sqrt(2.0))) {
            cond_1 = approx(len_1, 1.0) && approx(len_2, 1.0);
        }

        // Check triangle area according to (2).
        bool cond_2 = false;
        if (cond_1) {
            // Trianlge seems regular half 1-b-1 square. Just to be sure, check the area.
            cond_2 = approx(area, 0.5);
        }
        else {
            // Irregular triangle. Accept degenerate ones. (Assume they are on the outer egde, but do not check.)
            cond_2 = area < 0.01;
        }

        EXPECT_TRUE(cond_2) << "triangle = [" << p1.x << ", " << p1.y << "]; [" << p2.x << ", " << p2.y << "]; " << "[" << p3.x << ", " << p3.y << "]";
    }
}
```

**Comment by abellgithub** [comment]
Thanks for the great report with test.  As you may know, this is not my code and it doesn't surprise me that there are some issues.  I've just been trying to maintain it for my own use and for others who want to make use of it.  That said, I'll take a look at this when I can and see if I can figure out what's up.

**Comment by MK-3PP** [comment]
Thank you for forking and maintaining this branch. I'm aware of how this fork evolved and well understand your position. Seeing how this is the active C++-fork and how it differs from JS, I've choosen to post the issue here and not on mapbox. There are some issues with the tests against the JS-results, too, but I hadn't had the calm yet to rule out our local project setup as a cause.
Have to move on with the project for now, but I'll definitely check back into the issue with my colleagues in revision slots and share our findings, if they are helpful.

**Comment by abellgithub** [comment]
I've included your test in the suite and updated to use gtest.  When I run this, it passes.  I'm currently running on OSX.  Can you see if this [branch](https://github.com/abellgithub/delaunator-cpp/tree/gtest) works for you?  At this point you should be able to build with cmake and then run ctest in order to check.

**Comment by abellgithub** [comment]
Update - I tried more angles and got it to fail.  I'll pursue from here.

**Comment by abellgithub** [comment]
The degenerate triangles seem a problem to me, too.  They're made of three collinear points on the hull.  Thanks again for such a great test.

**Comment by MK-3PP** [comment]
I've tested the new branch and the issue persists at the given angle of 207°. I could not find any difference in the branch source code either, except from whitespace and commentary changes.
Output is
```
[path].cpp(108): error: Value of: cond_2
  Actual: false
Expected: true
triangle = [9.294073598055741, 14.270994619900026]; [7.495086074388281, 15.599017168537216]; [8.840083098316196, 15.162001144088393]
[path].cpp(108): error: Value of: cond_2
  Actual: false
Expected: true
triangle = [11.093061121723203, 12.942972071262835]; [9.3110480733464662, 12.034991071783743]; [9.7480640977952877, 13.379988095711656]
[path].cpp(108): error: Value of: cond_2
  Actual: false
Expected: true
triangle = [10.202054597534833, 12.48898157152329]; [7.495086074388281, 15.599017168537216]; [9.294073598055741, 14.270994619900026]
```

Or did you mean the changes you made around including GTest? With my current toolchain I am not able to automatically build and run with CMake. We got a classic MSVC setup built around Visual Studio, so I just copy the sources to our test projects. GTest was not mandatory for you to include. It is just what we use and I did not want to pseudo-code Catch with no backend to check against. Porting from GTest to Catch should be as simple as changing the TEST and EXPECT_TRUE macros.

**Comment by MK-3PP** [comment]
> 
> 
> The degenerate triangles seem a problem to me, too. They're made of three collinear points on the hull. 

To what extent? In our real world data they appear frequently, also spanning more than three consecutive collinear points (they are made of three collinear points on the hull, but "skip" a few points on their way).

It's a common artefact to be expected with that kind of algorithm. Usually they have to be filtered afterwards by constraining edge length, angles etc., that's why the test gracefully ignores them. Do you see a problem here that I overlooked?

**Comment by MK-3PP** [comment]
Ran the [branch](https://github.com/abellgithub/delaunator-cpp/tree/gtest) from fresh checkout. Built on gcc 7.3.0, 64-bit under Linux Mint. Results are the same as under Visual C++ 14.0.

It is run on the same machine in a VM, though. If this has something to do with rounding errors, maybe it depends on the CPU? Thin ice on my side here, but I hope we do not have to open the can of worms that is _Hardware- and OS-specific implementations of IEEE 754_.

```
saeba@saeba-Mint ~/delaunator_gtest $ file ./cmake-build/bin/issue-2
./cmake-build/bin/issue-2: ELF 64-bit LSB  executable, x86-64, version 1 (GNU/Linux), dynamically linked (uses shared libs), for GNU/Linux 2.6.24, BuildID[sha1]=8039133d068885e8ee42c07f1967170d6390b71d, not stripped

saeba@saeba-Mint ~/delaunator_gtest $ ./cmake-build/bin/issue-2
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from External_Lib_Delaunator_Test
[ RUN      ] External_Lib_Delaunator_Test.rotate_grid_6_8_8_207_0
/home/saeba/delaunator_gtest/test/issue-2.cpp:135: Failure
Value of: cond_2
  Actual: false
Expected: true
triangle = [9.294073598055741, 14.270994619900026]; [7.495086074388281, 15.599017168537216]; [8.840083098316196, 15.162001144088393]
/home/saeba/delaunator_gtest/test/issue-2.cpp:135: Failure
Value of: cond_2
  Actual: false
Expected: true
triangle = [11.093061121723203, 12.942972071262835]; [9.3110480733464662, 12.034991071783743]; [9.7480640977952877, 13.379988095711656]
/home/saeba/delaunator_gtest/test/issue-2.cpp:135: Failure
Value of: cond_2
  Actual: false
Expected: true
triangle = [10.202054597534833, 12.48898157152329]; [7.495086074388281, 15.599017168537216]; [9.294073598055741, 14.270994619900026]
[  FAILED  ] External_Lib_Delaunator_Test.rotate_grid_6_8_8_207_0 (0 ms)
[----------] 1 test from External_Lib_Delaunator_Test (0 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (0 ms total)
[  PASSED  ] 0 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] External_Lib_Delaunator_Test.rotate_grid_6_8_8_207_0

 1 FAILED TEST

```

**Comment by abellgithub** [comment]
I'll let you know when I know more.  I'm working it now.  It is strange that our systems yield different results.  Perhaps a function of different values for sin/cos that is used in your test rather than FP ops?  Even that seems a reach, but I'm not sure yet.

**Comment by abellgithub** [comment]
I seem to have things fixed, including the degenerate triangles, which seemed to be your ultimate problem.  I don't have an easy way to visualize things, though.  What did you use to construct the image in the ticket?

**Comment by abellgithub** [comment]
I take it back.  Still more work to do.

**Comment by MK-3PP** [comment]
I used Octave to plot the graphics since I already had it available. Octave uses gnuplot, but I do not know how to easily plot the mesh using only gnuplot. Matlab should work, too.

1. Print text file (C++)
```
inline void write_file(std::vector<Point2D> const& points, std::vector<size_t>& tris) {
    // Write octave format output file.
    // Use the following line in octave to visualize the result (Matlab could work, too). Replace $file with the actual file path:
    // clear; fid = fopen("$file"); eval(fgetl(fid)); eval(fgetl(fid)); eval(fgetl(fid)); triplot(tri, x, y); fclose(fid);
    char const* test_name = ::testing::UnitTest::GetInstance()->current_test_info()->name();
    ASSERT_NE(test_name, nullptr);
    std::ofstream ofs(std::string(test_name) + ".octave.txt");
    ASSERT_TRUE(ofs.is_open());

    ofs << "x = [";
    for (size_t i = 0; i < points.size(); ++i) {
        ofs << (i != 0 ? " ;" : "") << points[i].x;
    }
    ofs << "];" << std::endl;

    ofs << "y = [";
    for (size_t i = 0; i < points.size(); ++i) {
        ofs << (i != 0 ? " ;" : "") << points[i].y;
    }
    ofs << "];" << std::endl;

    ofs << "tri = [";
    for (size_t i = 0; i < tris.size(); i += 3) {
        ofs << (i != 0 ? " ;" : "");
        ofs << tris[i] + 1 << ", ";
        ofs << tris[i + 1] + 1 << ", ";
        ofs << tris[i + 2] + 1;
    }
    ofs << "];" << std::endl;
}
``` 
2. Visualize by entering this one-liner (Octave GUI)
 `clear; fid = fopen("path\\to\\foo.octave.txt"); eval(fgetl(fid)); eval(fgetl(fid)); eval(fgetl(fid)); triplot(tri, x, y); fclose(fid);`

**Comment by MK-3PP** [comment]
[I managed to pull the same thing off in JS](https://github.com/mapbox/delaunator/issues/61). Maybe the original creators can help.

**Comment by abellgithub** [comment]
I think I'll have it fixed shortly.  I fixed your case.

**Comment by abellgithub** [comment]
The problem is definitely determining collinearity (or, whether two vectors are clockwise or counter-clockwise).  The current code either chooses "clockwise" or "counterclockwise", but doesn't really deal with "neither".  I've modified the code to deal with this, but I'm trying to find the best way to determine whether three points should be considered collinear.  Mathematically, the determinant is 0 when points are collinear, but this doesn't work with finite-precision math.  People take various approaches with this.  Some try to do more precise math to get a determination.  I don't think this is the right approach for this library.  I think it makes more sense to use "close to 0" as indicating collinearity.  The question is how to determine what "close enough" is, taking into account that it's different based on the scale of the values.  When the numbers are large, the determinant is necessarily larger than when they are small, for the same relative layout of points.  I found a ratio that works in most cases, but there's still an issue that I'm trying to track down.

**Comment by abellgithub** [comment]
When two points are exceedingly close together (essentially the same), then things break down, because all measures relating the points break down.  What should happen to points that are, for all purposes, essentially identical?

**Comment by MK-3PP** [comment]
I feel I am not worked in enough (and won't be in the next days) to give a well-founded answer here. So please consider this an experience based gut instinct.

### [Collinearity and robustness](https://github.com/abellgithub/delaunator-cpp/issues/8#issuecomment-607883665)
Would it not be ok to have either CCW or CW, as long as it is consistent for any two vectors? So any set of points would have to be ordered by a criterion - and that must also have a way to be decisive. (Is that leading you to [your question regarding identical points](https://github.com/abellgithub/delaunator-cpp/issues/8#issuecomment-607895899)?)

I agree with you that increasing precision is not the right way, but rather to detect the corner cases and treat them as exception (let the triangulation fail, if there is no other way). Recently I tried detecting scale invariant precision issues by using the almost_equal example from [cppreference.com](https://en.cppreference.com/w/cpp/types/numeric_limits/epsilon):
```
template<class T>
typename std::enable_if<!std::numeric_limits<T>::is_integer, bool>::type
    almost_equal(T x, T y, int ulp)
{
    // the machine epsilon has to be scaled to the magnitude of the values used
    // and multiplied by the desired precision in ULPs (units in the last place)
    return std::fabs(x-y) <= std::numeric_limits<T>::epsilon() * std::fabs(x+y) * ulp
        // unless the result is subnormal
        || std::fabs(x-y) < std::numeric_limits<T>::min();
}
```
But while it gets the scaling problem (kinda? - I fail to overlook every FP precision gotcha) out of the way, it still remains dependent on a choice of sensible ULPs.
Thus I would prefer consistent decisive criterions.

@mourner [pointed out](https://github.com/mapbox/delaunator/issues/61#issuecomment-607902149) that robust predicates that could be used to make unambiguous choiches. I second that. He proposes [Shewchuk's](https://www.cs.cmu.edu/~quake/robust.html) predicates, but while these seem to be based on rock solid work, there is this disclaimer:

> **Warning:** This code will not work correctly on a processor that uses extended precision internal floating-point registers, such as the Intel 80486 and Pentium, unless the processor is configured to round internally to double precision. See the Predicates on PCs page for details.

Also the author urges you to read and understand his papers before using it.
Basically there initialization code must be called at runtime before using the predicates and one basically has to hand-tailor and sanity-check it against any platform the code should be built on.


### [Identical points](https://github.com/abellgithub/delaunator-cpp/issues/8#issuecomment-607895899)
Since 'close together' is a domain-dependent parameter, I have seen a bunch of Software (e.g. [GOM](https://www.gom.com/) that lets you enter a distance threshold that determines when points are to be treated as identical. What they do with it is usually not documented. Trivial cases would be
* choosing one point randomly / first in array
* hull (bounding box, sphere, ...) centroid
* centroid

**Comment by mourner** [comment]
> Warning: This code will not work correctly on a processor that uses extended precision internal floating-point registers, such as the Intel 80486 and Pentium, unless the processor is configured to round internally to double precision.

@MK-3PP this was written in 1996, and I don't think this holds true for current machines — the code should work properly. Also, there's an updated version of the code for modern times being maintained by @danshapero here: https://github.com/danshapero/predicates — I'd suggest trying it out.

**Comment by MK-3PP** [comment]
Sure, I will try it, thanks for pointing that out :)

But I do not think that the warning from 1996 has expired, since CPUs have become even more history-laden and abstract. AFAIK internal high res FP registers still are a thing and those are subject of Shewchuk's warning, because his code relies on a model of the underlying hardware. As far as I grasped it on superficial reading.

**Comment by abellgithub** [comment]
Shewchuk is wanting to make things mathematically correct to the extent possible with FP hardware.  But I'm not sure that determining whether some point (relatively) far away from some other points creates a left or right-handed system is that important to most users of this library.  I *think* what most people are wanting is to create a reasonable triangulation that they can feed to a graphics engine.  If, for example, we get a few triangles that don't meet the delaunay constraint in edge cases, that seems OK for most work.  It also seems OK if a few points are discarded because they are, relatively speaking, exceedingly close to some other point (the current code already does this).  What we clearly want to avoid is triangles that obviously overlap others and we want a correct convex hull up to points that are determined to be collinear based on whatever measure of this we choose.

I also have other criteria for this code.  1) it needs to perform reasonably well 2) it needs to be easily maintainable, because unlike with Shewchuk, delaunay triangulation is not my life, and as open source code, it's likely that it will get dropped by someone (me) and end up in the hands of another who needs to understand what's going on without getting a PhD.

Hopefully I'll be able to push a branch that addresses this issue shortly, though I'm not suggesting it will be perfect.

**Comment by danshapero** [comment]
@MK-3PP The real culprit was the x87 floating point units that were prevalent when Shewchuk was writing that code, which would internally use an 80-bit representation. Those processors are pretty rare now. The only scenario I can think of with modern hardware that might have the same behavior is using the FMA or fused multiply-add operation, which also uses higher precision on intermediate results. I haven't run into any cases where this made a difference but I would bet that there is one.

I'm happy to have other people use the updated predicates repo, but I haven't maintained it actively for a while. On top of that, I made several mistakes (since corrected) trying to update Shewchuk's code in the past, and it's possible there are other problems lurking.

**Comment by MK-3PP** [comment]
@abellgithub I totally appreciate the effort you put and are putting into this. And I agree with you, delaunator is extremely fast, comprehensible and self-contained. That is it's USP and it should stay that way. I do not want to force anything here. Just trying to answer every question you asked me and supplying data where possible.

My criteria stem from metrology, so they are a bit more towards robustness. But hey, for each job the right tool. Totally fine.

But - and that is just to make you aware of it - @Samfox2 and me are both having big point clouds (~1M points) and a few Triangles (~10 to 100) that span big portions of the point cloud's area. Since you wrote

> What we clearly want to avoid is triangles that obviously overlap others and we want a correct convex hull up to points that are determined to be collinear based on whatever measure of this we choose.

it is probably very good you looked into it and not in vain. I cannot show you my real world data because it shows a customer's CAD, but maybe @Samfox2 can?

_Point is: In small test cases we geht the fringes on the hull, like in the screenshot above. Big datasets do have overlapping triangles. Big overlapping triangles. As in:_
![Clipboard01](https://user-images.githubusercontent.com/60696116/78376801-1fc67000-75cf-11ea-96ff-d832b04a3b72.png)

Cheers and have a nice weekend!

**Comment by MK-3PP** [comment]
@danshapero Thank you very much for clarifying that.

**Comment by abellgithub** [comment]
If anyone would like to test this [branch](https://github.com/abellgithub/delaunator-cpp/tree/gtest), I'd appreciate it.  I'm not really done, but it seems to work better than what was there and handles the test case provided.  It seems to maintain points on the hull that were sometimes getting deleted before and eliminates "triangles" that were actually three collinear points.  Failure test cases appreciated.  There is a test failure, but it seems like a rounding problem, as it's off in the last decimal place of the precision available for double-precision FP, though I haven't yet examined the triangulation itself.

There's a lot of code change here, but most of it is just for my sanity as I try to figure this all out and doesn't really change any behavior.

**Comment by MK-3PP** [comment]
The robustness test failed as you mentioned, else the results are OK. Tested under Visual C++ 14.0 on Windows and gcc 7.3.0 on Linux Mint, both 64-bit:
```
saeba@saeba-Mint ~/delaunator_gtest/build $ ctest
Test project /home/saeba/delaunator_gtest/build
    Start 1: issue-2
1/2 Test #1: issue-2 ..........................   Passed    0.46 sec
    Start 2: delaunator.test
2/2 Test #2: delaunator.test ..................***Failed    0.01 sec

50% tests passed, 1 tests failed out of 2

Total Test time (real) =   0.48 sec

The following tests FAILED:
	  2 - delaunator.test (Failed)
Errors while running CTest
```
As a quick proof of concept, I _manually_ tested against some real world data with some minute changes: The data is a 3D-scan with numerically perfect collinear scan lines. I wiggled the normal vector of the projection plane (used to get 2D-data for delaunay triangulation) slightly, because that was what triggered our initial awareness of the bug, since it added numerical error to the the perfect collinearity of scan lines.

I tested two datasets with various X-Y-plane-ish projection planes.
_All manual tests were verified visually. They all turned out OK._
(images will be edited in from other machine)

I'd rather leave it up to you and the maintainers at [delaunator](https://github.com/mapbox/delaunator) to decide if and how the issue should be fixed, since the implications go beyond my understanding. I'll gladly help to test, though.
